### DISCUSSION TITLE
ESP32C3 with rule if/then construct

### DISCUSSION THREAD

**Discussion by refthoom** [arendst/Tasmota#21119] [Q&A] [answered]
Hi community. Let me start by saying that I'm truly amazed at what can be done with Tasmota 'out of the box'! Thanks so much to arendst and anyone else who has contributed.

For my project I'm using an 'ESP32C3 super mini'. Using the Tasmota web installer (just love it) I flashed it with, from the 'unofficial' section, 'Tasmota32 bluetooth (english)'. Attached hardware are two DS18B20, the PWM pin from a PC fan (to set the fan speed), an INA219 current sensor (I2C) and a mini DC motor driver (L298N/MX1508) using two GPIO's as PWM channels. A bunch of rules governs all triggers and actions such that with all that I receive a BT broadcast (external sensor), measure temperatures and adjust the fan speed from that, drive the DC motor and measure it’s DC current. The INA219 current sensor is used with a rule that sets both used GPIO low (PWM 0) if the current passes a threshold. Besides being a safety feature it is used to detect both end stops (fully open, fully closed) of the valve the DC motor is moving. Every action is sent to MQTT for my home automation.

In normal situations PWM4 and PWM5 are used to drive the DC motor, where PWM4 would be 1023 and PWM5 0 to turn it CW and vice versa to turn it CCW. On top of this I want to implement a safety feature that prevents PWM4 and PWM5 to have a value above 0 **at the same time** (as this would kill the L298N driver board).

So far I'm stuck building the appropriate rule for it. I've tried the On/Break/Endon method and a two days of trial and errors with if/endif in the rule but it just won't work. After each change I turned the rule on and off.

Last I tried was this (with and without brackets):
On Pwm#Pwm4>0 Do If (Pwm#Pwm4>0) And (Pwm#Pwm5>0) Backlog0 Pwm4 0; Pwm5 0 Endif Endon
This results in 'If:Done' when both channels are above zero but it does not execute the backlog part.

No results either with this:
ON Pwm4#State DO Var1=%value% ENDON
ON Pwm5#State DO Var2=%value% ENDON
ON Var1>0 Do If (Var1>0) And (Var2>0) Backlog0 Pwm4 0; Pwm5 0 Endif ENDON

I have to admit I don't fully understand the if/then construct inside rules. There aren't that many examples to find on this specific topic that give helpful clues, nor does the rules section or cookbook from the Tasmota documentation (which are otherwise excellent). Also, I looked at the Berry section but that clearly goes above my skill level.

Hopefully this explains my predicament and someone can give me a clue on what I'm missing or doing wrong here :-)

**Comment by sfromis** [comment]
Assuming that you already use `SetOption15 0`....

As most possible rule triggers are JSON-based, with a payload like this:
`{"PWM":{"PWM1":0,"PWM2":0,"PWM3":0,"PWM4":1023,"PWM5":0}}`
appropriate rule triggers would be `on PWM#PWM4 ...` and `on var2#state do ...`
You also need `()` around the *whole* condition in `if`, not just each comparison.

More generally, on ESP32 I'd recommend the Berry language, which is a lot less quirky, instead being more of a "normal" programming lkanguage.

**Comment by refthoom** [comment, marked as answer]
Hi @sfromis, thanks for the quick reply. 

Setoption 15 is 0 indeed, don't need the sliders etc. :-) I understand the JSON approach, therefor using pwm#pwm4 in the rule. Adding the extra ( ) did not solve it though.

`stat/tasmota_808E9C/RESULT = {"Rule1":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":87,"Free":424,"Rules":"On Pwm#Pwm4>0 Do If ((Pwm#Pwm4>0) And (Pwm#Pwm5>0)) Backlog0 Pwm4 0; Pwm5 0 Endif Endon"}}`

```
10:54:47.040 CMD: pwm4 10
10:54:47.044 MQT: stat/tasmota_808E9C/RESULT = {"PWM":{"PWM4":10,"PWM5":0}}
10:54:47.053 RUL: PWM#PWM4>0 performs "If ((Pwm#Pwm4>0) And (Pwm#Pwm5>0)) Backlog0 Pwm4 0; Pwm5 0 Endif"
10:54:47.057 MQT: stat/tasmota_808E9C/RESULT = {"If":"Done"}
10:55:01.669 CMD: pwm5 10
10:55:01.673 MQT: stat/tasmota_808E9C/RESULT = {"PWM":{"PWM4":10,"PWM5":10}}
10:55:01.682 RUL: PWM#PWM4>0 performs "If ((Pwm#Pwm4>0) And (Pwm#Pwm5>0)) Backlog0 Pwm4 0; Pwm5 0 Endif"
10:55:01.686 MQT: stat/tasmota_808E9C/RESULT = {"If":"Done"}
10:55:08.009 CMD: pwm
10:55:08.013 MQT: stat/tasmota_808E9C/RESULT = `{"PWM":{"PWM4":10,"PWM5":10}}
```
The trigger seems to be okay but the evaluation (still) seems to fail. Using pwm#pwm5 as trigger makes no difference, as expected.

Your other suggestion does work, sort of

`Rule1 ON Pwm#Pwm4 DO Var1=%value% ENDON ON Pwm#Pwm5 DO Var2=%value% ENDON ON var1#state Do If ((Var1>0) And (Var2>0)) Backlog0 Pwm4 0; Pwm5 0 Endif ENDON ON var2#state Do If ((Var2>0) And (Var1>0)) Backlog0 Pwm4 0; Pwm5 0 Endif ENDON`
results in this
```
11:17:42.878 CMD: pwm4 10
11:17:42.882 MQT: stat/tasmota_808E9C/RESULT = {"PWM":{"PWM4":10,"PWM5":0}}
11:17:42.888 RUL: PWM#PWM4 performs "Var1=10"
11:17:42.891 MQT: stat/tasmota_808E9C/RESULT = {"Var1":"10.000"}
11:17:42.897 RUL: PWM#PWM5 performs "Var2=0"
<cut>
11:18:22.726 CMD: pwm5 10
11:18:22.730 MQT: stat/tasmota_808E9C/RESULT = {"PWM":{"PWM4":10,"PWM5":10}}
<cut>
11:18:22.784 RUL: VAR1#STATE performs "If ((Var1>0) And (Var2>0)) Backlog0 Pwm4 0; Pwm5 0 Endif"
11:18:22.788 MQT: stat/tasmota_808E9C/RESULT = {"If":"Done"}
11:18:22.862 RUL: VAR2#STATE performs "If ((Var2>0) And (Var1>0)) Backlog0 Pwm4 0; Pwm5 0 Endif"
11:18:22.866 MQT: stat/tasmota_808E9C/RESULT = {"If":"Done"}
11:18:23.192 MQT: stat/tasmota_808E9C/RESULT = {"PWM":{"PWM4":10,"PWM5":0}}
11:18:23.199 RUL: PWM#PWM4 performs "Var1=10"
11:18:23.204 MQT: stat/tasmota_808E9C/RESULT = {"Var1":"10.000"}
11:18:23.210 RUL: PWM#PWM5 performs "Var2=0"
11:18:23.213 MQT: stat/tasmota_808E9C/RESULT = {"Var2":"0.000"}
11:18:23.225 MQT: stat/tasmota_808E9C/RESULT = {"PWM":{"PWM4":10,"PWM5":0}}
11:18:23.231 RUL: PWM#PWM4 performs "Var1=10"
11:18:23.235 MQT: stat/tasmota_808E9C/RESULT = {"Var1":"10.000"}
11:18:23.241 RUL: PWM#PWM5 performs "Var2=0"
11:18:23.245 MQT: stat/tasmota_808E9C/RESULT = {"Var2":"0.000"}
11:18:23.257 RUL: VAR1#STATE performs "If ((Var1>0) And (Var2>0)) Backlog0 Pwm4 0; Pwm5 0 Endif"
11:18:23.262 MQT: stat/tasmota_808E9C/RESULT = {"If":"Done"}
11:18:23.292 RUL: VAR2#STATE performs "If ((Var2>0) And (Var1>0)) Backlog0 Pwm4 0; Pwm5 0 Endif"
11:18:23.297 MQT: stat/tasmota_808E9C/RESULT = {"If":"Done"}
11:18:23.497 MQT: stat/tasmota_808E9C/RESULT = {"PWM":{"PWM4":0,"PWM5":0}}
11:18:23.504 RUL: PWM#PWM4 performs "Var1=0"
11:18:23.508 MQT: stat/tasmota_808E9C/RESULT = {"Var1":"0.000"}
11:18:23.514 RUL: PWM#PWM5 performs "Var2=0"
11:18:23.518 MQT: stat/tasmota_808E9C/RESULT = {"Var2":"0.000"}
```
It seems there is some 'looping' going on before the evaluation executes? I wil test this further with a breadboard as above results are only software and then come back here later. Thanks a lot so far!

**Reply by sfromis** [reply]
You can only use trigger expressions with `#` in the trigger part of the rule, between `on` and `do`, not anywhere else. It is **nothing** like a programming language where you can expect constructs to work where they would make sense.

When it comes to your last attempt there, you should **only** have the rule for `var2#state`, as both var1 and var2 should be set at the same time, and therefore you should not trigger on var1. However, it does not work well when you pick up the result again.

What I posted was about the syntax, not about making the logic work in rules. They are indeed quirky to use, hence my recommendation of using Berry instead.

**Reply by refthoom** [reply]
So, I found a solution. Although I didn't have any progress with berry and rethinking what I need, at some point I realised I was making things too complicated: Adding this to the rule 

> on pwm#pwm4>0 do pwm5 0 endon on pwm#pwm5>0 do pwm4 0 endon

actually provides the safety measure I was looking for.

Thanks for your time and effort @sfromis :-)
